### PR TITLE
feat: Implemented all developer functions

### PR DESCRIPTION
**Pull request by AkshatGKMit**

### CODE REVIEW

**Comment by adiTrivsme** [comment]
Screenshot 2024-09-27 at 9 23 47 AM

The commit message is not clear, fine for now but keep messages clear moving forward.

The message could have been:

`feat: add remove developer function and the callbacks for remove condition`

**Comment by adiTrivsme** [comment]
Screenshot 2024-09-27 at 9 27 06 AM

make sure you don't make typos in commit messages if done by mistake then edit the commit before pushing to remote

File: src/scripts/developer.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { validateDeveloper, generateId } from "./helpers/index.js";
+
+export function addDeveloper(developers: Developer[], developer: Partial<Developer>) {
+	const isDevValidated = validateDeveloper(developer);
+	if (isDevValidated) {
+		const validatedDev: Developer = {
+			...developer,
+			id: generateId(developers),
+		} as Developer;
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: why using `as` if the type `Developer` is already given on variable

File: src/scripts/developer.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		developers.push(validatedDev);
+
+		alert(`Developer ${developer.name} added`);
+	}
+}
+
+export function cloneDeveloper(developers: Developer[], developer: Developer): Developer {
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: use type alias to define type for `developers`

File: src/scripts/developer.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	const deepCopy: Developer = {
+		...JSON.parse(JSON.stringify(developer)),
+		id: generateId(developers),
+	};
+	return deepCopy;
+}
+
+export function updateDeveloper(developer: Developer, updates: Partial<Developer>) {
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: 1. do not directly modify `developer`, in most cases function should not directly modify its arguments, clone the argument, update and return 
2. code can be refactored, this works but in modern JS we have better and cleaner options than this

File: src/scripts/developer.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	developer.name = updates.name ?? name;
+	developer.age = updates.age ?? age;
+	developer.experience = updates.experience ?? experience;
+	developer.isEmployed = updates.isEmployed ?? isEmployed;
+	developer.projects = updates.projects ?? projects;
+	developer.skills = updates.skills ?? skills;
+}
+
+export function findDevelopersBySkill(developers: Developer[], matchSkill: string): Developer[] {
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: use a type alias for `Developer[]` and use it everywhere

File: src/scripts/developer.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		alert(`'${newSkill}' is already present in developer '${name}'`);
+		return;
+	}
+
+	skills.push(newSkill);
+	alert(`'${newSkill}' is added to developer '${name}'`);
+}
+
+export function updateSkill(developer: Developer, oldSkill: string, newSkill: string) {
+	const { name, skills } = developer;
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: do not modify the developer directly

File: src/scripts/developer.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		alert(`'${oldSkill}' is already present in developer '${name}'`);
+		return;
+	}
+
+	alert(`'${oldSkill}' is updated to '${newSkill}' for developer ${name}`);
+	skills[oldSkillIndex] = newSkill;
+}
+
+export function removeDeveloperByCondition({ devs, condition }: RemoveDeveloperByConditionType): Developer[] {
+	return devs.filter((developer) => !condition(developer));
```
Comment (adiTrivsme) [inline src/scripts/developer.ts]: `condition` variable name is not intuitive, its not signifying that its a callback

File: src/scripts/helpers/index.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
+export function validateDeveloper(dev: Partial<Developer>): boolean {
+	const { name, age, isEmployed, experience, skills } = dev;
+
+	if (!name) {
+		alert("Invalid developer's name");
+		return false;
+	}
+
+	if (name === "") {
```
Comment (adiTrivsme) [inline src/scripts/helpers/index.ts]: redundant code

File: src/scripts/helpers/index.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		return false;
+	}
+
+	if (age < 10) {
+		alert("Developer age cannot be less than 10");
+		return false;
+	}
+
+	if (age > 54) {
+		alert("Developer age cannot be more than 54s");
```
Comment (adiTrivsme) [inline src/scripts/helpers/index.ts]: what is `s` in 54

File: src/scripts/helpers/index.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+export function removeDevsByAge({ age }: Developer): boolean {
+	return age < 15;
+}
+
+export function removeDevsByIncompleteProject({ projects }: Developer): boolean {
+	return projects.reduce((total, { isCompleted }) => (total += isCompleted ? 0 : 1), 0) > 1;
+}
+
+export function removeDevsBySkill({ skills }: Developer): boolean {
```
Comment (adiTrivsme) [inline src/scripts/helpers/index.ts]: make `skill` dynamic - when using the function one should be able to define by what skill he wants to remove the developer

File: src/scripts/helpers/index.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	return skills.map((skill) => skill.toLowerCase()).indexOf("git") === -1;
+}
+
+export function removeDevsByExperienceAndProjects({ experience, projects }: Developer): boolean {
+	return experience >= 3 && projects.length < 2;
```
Comment (adiTrivsme) [inline src/scripts/helpers/index.ts]: same here make the condition dynamic

**Review by adiTrivsme — CHANGES_REQUESTED** [review]

File: src/scripts/developer.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	};
+	developers.push(validatedDev as Developer);
+
+	alert(`Developer ${developer.name} added`);
+}
+
+export function cloneDeveloper(developer: Developer, developers?: Developers): Developer {
+	const deepCopy: Developer = {
+		...JSON.parse(JSON.stringify(developer)),
+		id: developers !== undefined ? generateId(developers) : developer.id,
```
Comment (shelly19gkmit) [inline src/scripts/developer.ts]: shorthand code- replace developers !== undefined  with !!developers;
and 
Destructure id

File: src/scripts/developer.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	};
+
+	return deepCopy;
+}
+
+export function updateDeveloper(developer: Developer, updates: Partial<Developer>): Developer {
+	const clonedDev = cloneDeveloper(developer);
+	const { name, age, experience, isEmployed, projects, skills } = clonedDev;
```
Comment (shelly19gkmit) [inline src/scripts/developer.ts]: Add null check

File: src/scripts/developer.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	return cloneDev;
+}
+
+export function removeDeveloperByCondition({ devs, callbackCondition, args = [] }: RemoveDeveloperByConditionType): Developers {
+	return devs.filter((developer) => !callbackCondition(developer, ...args));
+}
+
+export function sortDevelopersByEmploymentAndAge(developers: Developers, sortByAgeAscending: boolean): Developers {
+	const sortedDevs = developers.sort((aDev: Developer, bDev: Developer) => {
+		if (aDev.isEmployed && bDev.isEmployed) {
```
Comment (shelly19gkmit) [inline src/scripts/developer.ts]: What is aDev and bDev? Rename to more understandable name

File: src/scripts/developer.ts (hunk 12 above)
Comment (AkshatGKMit) [reply src/scripts/developer.ts]: where i have to add null check in this

Comment (shelly19gkmit) [reply src/scripts/developer.ts]: Like this -> const { name, age, experience, isEmployed, projects, skills } = clonedDev || {};

File: src/scripts/helpers/index.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+export function removeDevsBySkill({ skills }: Developer, skill: string): boolean {
+	if (typeof skill !== "string") skill = "react";
+	return skills.map((skill) => skill.toLowerCase()).indexOf(skill.toLowerCase()) !== -1;
+}
+
+export function removeDevsByExperienceAndProjects({ experience, projects }: Developer, minExperience: number, projectLength: number): boolean {
+	if (!minExperience || typeof minExperience !== "number") minExperience = 2;
+	if (!projectLength || typeof projectLength !== "number") projectLength = 1;
+	return experience >= minExperience && projects.length < projectLength;
+}
```
Comment (shelly19gkmit) [inline src/scripts/helpers/index.ts]: create a reusable function to check if value is number or not

File: src/scripts/developer.ts (hunk 12 above)
Comment (AkshatGKMit) [reply src/scripts/developer.ts]: ok

**Review by shelly19gkmit — APPROVED** [review]

**Pull request merged**